stdlib/src/array.rs

```diff
@@ -1077,6 +1109,52 @@ mod array {
             }
             Ok(true)
         }
+
+        #[pymethod(magic)]
+        fn reduce_ex(
+            zelf: PyRef<Self>,
+            proto: usize,
+            vm: &VirtualMachine,
+        ) -> PyResult<(PyObjectRef, PyObjectRef, Option<PyDictRef>)> {
+            if proto < 3 {
+                return Self::reduce(zelf, vm);
+            }
+            let array = zelf.read();
+            let cls = zelf.as_object().clone_class().into_object();
+            let typecode = vm.ctx.new_utf8_str(array.typecode_str());
+            let bytes = vm.ctx.new_bytes(array.get_bytes().to_vec());
+            let code = MachineFormatCode::from_typecode(array.typecode()).unwrap();
+            let code = PyInt::from(u8::from(code)).into_object(vm);
+            let module = vm.import("array", None, 0)?;
+            let func = vm.get_attribute(module, "_array_reconstructor")?;
+            Ok((
+                func,
+                vm.ctx.new_tuple(vec![cls, typecode, code, bytes]),
+                zelf.as_object().dict(),
+            ))
+        }
+
+        #[pymethod(magic)]
+        fn reduce(
+            zelf: PyRef<Self>,
+            vm: &VirtualMachine,
+        ) -> PyResult<(PyObjectRef, PyObjectRef, Option<PyDictRef>)> {
+            let array = zelf.read();
+            let cls = zelf.as_object().clone_class().into_object();
+            let typecode = vm.ctx.new_utf8_str(array.typecode_str());
+            let values = if array.typecode() == 'u' {
+                let s = Self::_wchar_bytes_to_string(array.get_bytes(), array.itemsize(), vm)?;
+                s.chars().map(|x| x.into_pyobject(vm)).collect()
```

Comment:
In windows, I tried this code with CPython:
```python
test_str = '🌉abc🌐def🌉🌐'
u = array('u', test_str)
print(u.__reduce_ex__(1))
```
then
```
  File ".\extra_tests\snippets\stdlib_array.py", line 104, in <module>
    assert u.__reduce_ex__(1)[1][1] == list(test_str), (u.__reduce_ex__(1)[1][1], list(test_str))
AssertionError: (['\ud83c', '\udf09', 'a', 'b', 'c', '\ud83c', '\udf10', 'd', 'e', 'f', '\ud83c', '\udf09', '\ud83c', '\udf10'], ['🌉', 'a', 'b', 'c', '🌐', 'd', 'e', 'f', '🌉', '🌐'])
```
The values are wchar encoded in CPython but doesn't look like that here

Reply:
Do you mean CPython did not treat it as same in linux macos?

Reply:
I don't mind if you want to just skip testing for 16bit character environments.